### PR TITLE
changes to support container environments

### PR DESCRIPTION
**Pull request by jedwards4b**
Support for running cesm in a container environment.   I have provided an example which runs in 
a cray environment container on cheyenne.  Please see [ESCOMP/CESM/wiki/Using-CESM-in-the-Cray-environment-container-on-Cheyenne](https://github.com/ESCOMP/CESM/wiki/Using-CESM-in-the-Cray-environment-container-on-Cheyenne) for usage details.

Test suite: scripts_regression_tests.py on cheyenne - lots of hand testing in crayenv
Test baseline:
Test namelist changes:
Test status: bit for bit

Fixes 

User interface changes?:

Update gh-pages html (Y/N)?:

### CODE REVIEW

**Comment by rljacob** [comment]
Why so many changes?   All we did was make a machine definition for the container.   No changes to CIME itself needed.
https://acme-climate.atlassian.net/wiki/spaces/DOC/pages/1429831693/Singularity+container

**Comment by jedwards4b** [comment]
@rljacob With these changes the container is pretty much invisible to the user, they simple go to the case directory and interact with cime in the same way that they would outside a container.

**Comment by rljacob** [comment]
How does this work for the user?  I looked for the example but couldn't find it.

**Comment by jedwards4b** [comment]
@rljacob I added a link to the usage in the initial comment.

**Comment by rljacob** [comment]
Oh there it is. What does the "crayenv" command do?  Is that something only on cheyenne?

File: scripts/Tools/case.build [hunk 1]
```diff
@@ -146,6 +149,8 @@ def parse_command_line(args, description):
         "files in the source tree or in SourceMods.",
     )
 
+
```
Comment (jgfouca) [inline scripts/Tools/case.build:152]: These whitespaces aren't needed.

**Comment by jedwards4b** [comment]
crayenv is a script that runs the singularity container on cheyenne which emulates the cray environment of 
our future system derecho.   It's specific to cheyenne and developers will need to create a similar script for
singularity on other systems.

File: scripts/lib/CIME/case/case.py [hunk 2]
```diff
@@ -163,10 +163,13 @@ def __init__(self, case_root=None, read_only=True, record=False, non_local=False
         # check if case has been configured and if so initialize derived
         if self.get_value("CASEROOT") is not None:
             if not self._non_local:
+                container = self.get_value("CONTAINER_ENVIRONMENT")
                 mach = self.get_value("MACH")
                 machobj = Machines()
                 probed_machine = machobj.probe_machine_name()
                 if probed_machine:
+                    if mach != probed_machine:
+                        return
```
Comment (jgfouca) [inline scripts/lib/CIME/case/case.py:171]: We can get rid of the expect below.

**Review by jgfouca — CHANGES_REQUESTED** [review]
Looks fine. Couple minor changes requested.

**Comment by rljacob** [comment]
Can you point me to that script?

**Comment by jedwards4b** [comment]
crayenv version 22.02
<details>
<summary>Details</summary>

#!/bin/bash
#
# This script expects 1 environment variable
#   1. SENV_IMAGEROOT - The root of Image Sandbox for Singularity
#
# -- Need to get rid of this in production
export SENV_IMAGEROOT=/glade/u/apps/ch/opt/hpe-cpe/22.02/cray_hpe_cpe_22.02-021622

ALLARGS=" $@ "
NARGS=$#
me=$(whoami)
THESHELL="/bin/bash"

LOGFIL="/glade/scratch/csgteam/logs/cpe/${me}.$(hostname -s).$$"
echo "Started: $(date +'%Y-%m-%dT%H:%M:%S')" > ${LOGFIL}

source /etc/profile.d/modules.sh
ml singularity
SCRDIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
export SINGULARITY=$(which singularity)
export SCR_IMAGEROOT=${SENV_IMAGEROOT}
export SINGULARITYENV_OMP_NUM_THREADS=$OMP_NUM_THREADS
export STARTUPENV=${SCRDIR}/startupenv
export STARTUPRC=${SCRDIR}/bashenv
export SINGULARITYENV_BASH_ENV=$STARTUPRC
export TERM=vt100

if [ -n "$CRAYENV_GPU_SUPPORT" ]; then
    export GPUARGS="--nv"
else
    export GPUARGS=
fi

export BINDARGS="-B/glade,/var/spool/pbs,/var/run/munge,/glade/u/apps/opt/nvhpc/21.3:/opt/nvidia/hpc_sdk,/glade/u/apps/ch/opt/hpe-cpe/22.02/ncar:/usr/ncar"

if [ ! -z "$PBS_JOBID" ] ; then
    export JUST_JOBID=$(echo $PBS_JOBID|awk -F\. '{print $1}')
    DEB_SCRATCH=/glade/scratch/${me}/.palsd/${JUST_JOBID}/var/tmp/$(hostname -s)
    PALSETC=/glade/scratch/${me}/.palsd/${JUST_JOBID}/etc
    if [ ! -d ${PALSETC} ] ; then
       mkdir -p ${PALSETC}
    fi
    if [ ! -d ${DEB_SCRATCH} ] ; then
       mkdir -p ${DEB_SCRATCH}
    fi
    LOC_BINDARGS="${BINDARGS},${DEB_SCRATCH}:/tmp,${DEB_SCRATCH}:/var/run/palsd,${PALSETC}:/etc/pals"
#   ${SCRDIR}/start_slurm.sh < /dev/null
    export DEB_SCRATCH
    export PALSD_LOGFILE="/tmp/palsd.log"
    export PALSD_DEBUG=2
    export PALSD_RUN_DIR="/tmp/palsd"
    PALSPORT=$(( 7919 + $(( $JUST_JOBID % 8192 ))))
    export PALSD_PORT=${PALSPORT}
    export PALS_PORT=${PALSPORT}
    export APRUN_PORT=${PALSPORT}
    export RFE_811452_DISABLE=1
    ${SCRDIR}/start_palsdaemons.sh < /dev/null
else
    DEB_SCRATCH=/glade/scratch/${me}
    LOC_BINDARGS="${BINDARGS},${DEB_SCRATCH}:/tmp"
fi

shargs="--rcfile ${STARTUPRC}"

ml purge
unset MODULEPATH_ROOT
unset MODULEPATH
unset MODULESHOME
unset -f module
unset -f ml

if [ $NARGS -eq 0 ] ; then
   ${SINGULARITY} run -u ${LOC_BINDARGS} ${GPUARGS} --env-file ${STARTUPENV} \
              ${SCR_IMAGEROOT} ${THESHELL} ${shargs}
else
   ${SINGULARITY} run -u ${LOC_BINDARGS} ${GPUARGS} --env-file ${STARTUPENV} \
              ${SCR_IMAGEROOT} ${THESHELL} -c "${ALLARGS}"
fi
echo "Ended: $(date +'%Y-%m-%dT%H:%M:%S')" >> ${LOGFIL}


</details>

**Comment by rljacob** [comment]
Thanks. is "ml" the same as "module load" ?

**Comment by jedwards4b** [comment]
yes

**Comment by rljacob** [comment]
So the command "crayenv ./create_newcase --case foo --compset X --res f19_g17" will start the container, execute the create_newcase command and leave you in the container for the remaining commands (qsub and ./case.submit).

**Comment by jedwards4b** [comment]
No it does not leave you in the container.  Each command (case.setup, case.build, case.submit) looks to see if the case expects a container and if it does the command reinvokes itself with the container prepended to the original command.   [Here is an example](https://github.com/ESMCI/cime/pull/4201/files#diff-015eba23e5899a0cd42e586b590fb93b89a4952470266044bd874386daebfa30R207)

File: config/xml_schemas/config_machines.xsd [hunk 3]
```diff
@@ -75,8 +76,8 @@
   <xs:complexType name="AttrElement">
     <xs:simpleContent>
       <xs:extension base="xs:string">
-        <xs:attribute ref="compiler">
-        </xs:attribute>
+	<xs:attribute ref="compiler">
```
Comment (rljacob) [inline config/xml_schemas/config_machines.xsd]: Are these white space changed intended?

**Comment by rljacob** [comment]
Ok.  Is CONTAINER_ENVIRONMENT in the machine file supposed to be set to the script name?  crayenv for cheyenne?

**Comment by jedwards4b** [comment]
Correct.

**Comment by jedwards4b** [comment]
superseded by #4238

**Pull request closed without merging**